src/contacts/ContactImporter.ts

```diff
@@ -0,0 +1,165 @@
+import m, { Children } from "mithril"
+import { List, ListAttrs, ListLoadingState, MultiselectMode, ViewHolder } from "../gui/base/List.js"
+import { px, size } from "../gui/size.js"
+import { VirtualRow } from "../gui/base/ListUtils.js"
+import { getContactListName } from "./model/ContactUtils.js"
+
+export async function importContactsFromFile(vCardData: string[], contactListId: string) {
+	const contactMembership = assertNotNull(locator.logins.getUserController().user.memberships.find((m) => m.groupType === GroupType.Contact))
+	const contacts = vCardListToContacts(vCardData, contactMembership.group)
+
+	return contactImportDialog(contacts, (dialog) => {
+		dialog.close()
+		importContacts(contacts, contactListId, locator.contactFacade)
+	}).show()
+}
+
+async function importContacts(contacts: ReadonlyArray<Contact>, contactListId: string, contactFacade: ContactFacade) {
+	const importPromise = contactFacade
+		.importContactList(contacts, contactListId)
+		.catch(
+			ofClass(ImportError, (e) =>
+				Dialog.message(() =>
+					lang.get("importContactsError_msg", {
+						"{amount}": e.numFailed + "",
+						"{total}": contacts.length + "",
+					}),
+				),
+			),
+		)
+		.catch(() => Dialog.message("unknownError_msg"))
+	await showProgressDialog("pleaseWait_msg", importPromise)
+	await Dialog.message(() =>
+		lang.get("importVCardSuccess_msg", {
+			"{1}": contacts.length,
+		}),
+	)
+}
+
+/**
+ * Show a dialog with a preview of a given list of contacts
+ * @param contacts The contact list to be previewed
+ * @param okAction The action to be executed when the user press the import button
+ */
+function contactImportDialog(contacts: Contact[], okAction: (dialog: Dialog) => unknown): Dialog {
+	return Dialog.showActionDialog({
+		title: () => lang.get("importVCard_action"),
+		child: () =>
+			m(
+				".mt-s",
+				m(
+					".flex.col.rel",
+					{
+						style: {
+							height: px(size.list_row_height * contacts.length),
+						},
+					},
+					m(List, {
+						renderConfig: {
+							itemHeight: size.list_row_height,
+							multiselectionAllowed: MultiselectMode.Disabled,
+							swipe: null,
+							createElement: (dom) => {
+								const row: ImportContactRowHolder = new ImportContactRowHolder(dom)
+
+								m.render(dom, row.render())
+
+								return row
+							},
+						},
+						state: {
+							items: contacts,
+							loadingStatus: ListLoadingState.Done,
+							loadingAll: false,
+							selectedItems: new Set(),
+							inMultiselect: false,
+							activeIndex: null,
+						},
+						onLoadMore() {},
+						onRangeSelectionTowards(item: Contact) {},
+						onRetryLoading() {},
+						onSingleSelection(item: any) {},
+						onSingleTogglingMultiselection(item: Contact) {},
+						onStopLoading() {},
+					} satisfies ListAttrs<Contact, ImportContactRowHolder>),
+				),
+			),
+		okAction,
+		allowCancel: true,
+		okActionTextId: "import_action",
+	})
+}
+
+class ImportContactVirtualRow implements VirtualRow<Contact> {
```

Comment:
I know you are following an existing pattern but in this case we can collapse both row and viewHolder into viewHolder